src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs

```diff
@@ -269,24 +269,31 @@ internal int ReadCore(Span<byte> buffer)
                     break;
                 }
 
-                if (_inflater.Finished())
+                // If the stream is finished then we have a few potential cases here:
+                // 1. DeflateStream => return
+                // 2. GZipStream that is finished but may have an additional GZipStream appended => feed more input
+                // 3. GZipStream that is finished and appended with garbage => return
+                if (_inflater.Finished() && (!_inflater.IsGzipStream() || !_inflater.NeedsInput()))
                 {
                     break;
                 }
 
-                int bytes = _stream.Read(_buffer, 0, _buffer.Length);
-                if (bytes <= 0)
+                if (_inflater.NeedsInput())
```

Comment:
Is NeedsInput cheap?  Just wondering if we called it above if there's any benefit in caching that we called it so we don't need to call it again.

Reply:
It's very cheap, yes. It just checks if a variable is set to zero.

Reply:
Ok